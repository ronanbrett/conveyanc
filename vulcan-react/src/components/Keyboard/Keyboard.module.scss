@import "src/styles/vars";

.Keyboard {

}