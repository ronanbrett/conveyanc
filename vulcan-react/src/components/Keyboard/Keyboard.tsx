import React, {
  Children,
  cloneElement,
  FC,
  KeyboardEvent,
  useCallback,
  useEffect,
} from "react";

const KEYS: { [id: number]: string } = {
  8: "onBackspace",
  9: "onTab",
  13: "onEnter",
  27: "onEsc",
  32: "onSpace",
  37: "onLeft",
  38: "onUp",
  39: "onRight",
  40: "onDown",
  188: "onComma",
  16: "onShift",
};

interface KeyboardProps {
  children?: any;
  onKeyDown?: (event: KeyboardEvent, ...args: any) => void;
  target?: string;
  [id: string]: any;
}

const Keyboard = ({ target, children, onKeyDown, ...props }: KeyboardProps) => {
  const onKeyDownHandler = useCallback(
    (event, ...rest) => {
      const key: number = event.keyCode ? event.keyCode : event.which;
      const callbackName = KEYS[key];

      if (callbackName && props[callbackName]) {
        props[callbackName](event, ...rest);
      }

      if (onKeyDown) {
        onKeyDown(event, ...rest);
      }
    },
    [onKeyDown, props]
  );

  useEffect(() => {
    if (target === "document") {
      document.addEventListener("keydown", onKeyDownHandler);
    }

    return () => {
      if (target === "document") {
        document.removeEventListener("keydown", onKeyDownHandler);
      }
    };
  }, [onKeyDownHandler, target]);

  return target === "document"
    ? children
    : cloneElement(Children.only(children), {
        onKeyDown: onKeyDownHandler,
      });
};

export default Keyboard;
