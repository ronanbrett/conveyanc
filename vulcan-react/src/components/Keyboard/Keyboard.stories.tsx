import React from "react";

import Keyboard from "./Keyboard";

import { action } from "@storybook/addon-actions";

export default {
  title: "Components/Keyboard",
  decorators: [],
  component: Keyboard,
  argTypes: { onEsc: { action: "clicked" } },
};

export const Default = (props) => (
  <Keyboard>
    <input type="text" />
  </Keyboard>
);
