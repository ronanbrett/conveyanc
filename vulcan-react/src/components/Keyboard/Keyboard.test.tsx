import React from "react";
import { render } from "@testing-library/react";
import Keyboard from "./Keyboard";

describe("Keyboard", () => {
  it("renders without error", () => {
    const { asFragment } = render(
      <Keyboard>
        <p></p>
      </Keyboard>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
