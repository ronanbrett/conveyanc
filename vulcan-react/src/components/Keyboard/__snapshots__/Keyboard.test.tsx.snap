// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard renders without error 1`] = `
<DocumentFragment>
  <p />
</DocumentFragment>
`;
